x/cronos/keeper/ibc.go

```diff
@@ -0,0 +1,157 @@
+package keeper
+
+import (
+	"errors"
+
+	"github.com/armon/go-metrics"
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	ibctransfertypes "github.com/cosmos/ibc-go/modules/apps/transfer/types"
+	ibcclienttypes "github.com/cosmos/ibc-go/modules/core/02-client/types"
+	"github.com/crypto-org-chain/cronos/x/cronos/types"
+)
+
+func (k Keeper) ConvertVouchersToEvmCoins(ctx sdk.Context, from string, coins sdk.Coins) error {
+	acc, err := sdk.AccAddressFromBech32(from)
+	if err != nil {
+		return err
+	}
+
+	params := k.GetParams(ctx)
+	evmParams := k.GetEvmParams(ctx)
+	for _, c := range coins {
+		switch c.Denom {
+		case params.IbcCroDenom:
+			if params.IbcCroDenom == "" {
+				return sdkerrors.Wrap(types.ErrIbcCroDenomEmpty, "ibc is disabled")
+			}
+
+			// Send ibc tokens to escrow address
+			err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, acc, types.ModuleName, sdk.NewCoins(c))
+			if err != nil {
+				return err
+			}
+			// Compute new amount, because basecro is a 8 decimals token, we need to multiply by 10^10 to make it
+			// a 18 decimals token
+			amount18dec := sdk.NewCoin(evmParams.EvmDenom, c.Amount.Mul(sdk.NewIntFromBigInt(types.TenPowTen)))
+
+			// Mint new evm tokens
+			if err := k.bankKeeper.MintCoins(
+				ctx, types.ModuleName, sdk.NewCoins(amount18dec),
+			); err != nil {
+				return err
+			}
+
+			// Send evm tokens to receiver
+			if err := k.bankKeeper.SendCoinsFromModuleToAccount(
+				ctx, types.ModuleName, acc, sdk.NewCoins(amount18dec),
+			); err != nil {
+				return err
+			}
+
+		default:
+			// TODO handle ERC20 tokens
+		}
+	}
+	defer func() {
+		for _, a := range coins {
+			if a.Amount.IsInt64() {
+				telemetry.SetGaugeWithLabels(
+					[]string{"tx", "msg", "ConvertVouchersToEvmCoins"},
+					float32(a.Amount.Int64()),
+					[]metrics.Label{telemetry.NewLabel("denom", a.Denom)},
+				)
+			}
+		}
+	}()
+	return nil
+}
+
+func (k Keeper) IbcTransferCoins(ctx sdk.Context, from, destination string, coins sdk.Coins) error {
+	acc, err := sdk.AccAddressFromBech32(from)
+	if err != nil {
+		return err
+	}
+
+	if len(destination) == 0 {
+		return errors.New("to address cannot be empty")
+	}
+
+	params := k.GetParams(ctx)
+	evmParams := k.GetEvmParams(ctx)
+
+	for _, c := range coins {
+		switch c.Denom {
+		case evmParams.EvmDenom:
+			// Compute the remainder, we won't transfer anything lower than 10^10
+			amount8decRem := c.Amount.Mod(sdk.NewIntFromBigInt(types.TenPowTen))
+			amountToBurn := c.Amount.Sub(amount8decRem)
+			if amountToBurn.IsZero() {
+				// Amount too small
+				break
```

Comment:
why break here?

Reply:
better to fail here?

I thing msg_send_coins sanitise by removing zero amount coin but does not fail the whole transaction

Reply:
If the evm denom coin has amount smaller than 0.000000001, should it be treated as a valid case and not prevent other coins from processing?

Reply:
yes, it should be continue..
fixed